File: lib/haveapi/authorization.rb
module HaveAPI
  class Authorization
    def initialize(&block)
      @block = block
    end

    # Returns true if user is authorized.
    # Block must call allow to authorize user, default rule is deny.
    def authorized?(user)
      @restrict = []

      catch(:rule) do
        instance_exec(user, &@block) if @block
        deny # will not be called if block throws allow
      end
    end

    # Apply restrictions on query which selects objects from database.
    # Most common usage is restrict user to access only objects he owns.
    def restrict(*args)
      @restrict << args.first
    end

    # Restrict parameters client can set/change.
    # [whitelist]  allow only listed parameters
    # [blacklist]  allow all parameters except listed ones
    def input(whitelist: nil, blacklist: nil)
      @input = {
          whitelist: whitelist,
          blacklist: blacklist,
      }
    end

    # Restrict parameters client can retrieve.
    # [whitelist]  allow only listed parameters
    # [blacklist]  allow all parameters except listed ones
    def output(whitelist: nil, blacklist: nil)
      @output = {
          whitelist: whitelist,
          blacklist: blacklist,
      }
    end

    def allow
      throw(:rule, true)
    end

    def deny
      throw(:rule, false)
    end

    def restrictions
      ret = {}

      @restrict.each do |r|
        ret.update(r)
      end

      ret
    end

    def filter_input(input, params)
      filter_inner(input, @input, params, false)
    end

    def filter_output(output, params, format = false)
      filter_inner(output, @output, params, format)
    end

    private
    def filter_inner(allowed_params, direction, params, format)
      allowed = {}

      allowed_params.each do |p|
        if params.has_param?(p.name)
          allowed[p.name] = format ? p.format_output(params[p.name]) : params[p.name]

        elsif params.has_param?(p.name.to_s) # FIXME: remove double checking
          allowed[p.name] = format ? p.format_output(params[p.name.to_s]) : params[p.name.to_s]
        end
      end

      return allowed unless direction

      if direction[:whitelist]
        ret = {}

        direction[:whitelist].each do |p|
          ret[p] = allowed[p] if allowed.has_key?(p)
        end

        ret

      elsif direction[:blacklist]
        ret = allowed.dup

        direction[:blacklist].each do |p|
          ret.delete(p)
        end

        ret

      else
        allowed
      end
    end
  end
end

File: lib/haveapi/authorization.ex
defmodule HaveAPI.Authorization do
  def authorize(%HaveAPI.Request{} = req) do
    ret = do_authorize(
      param_names(req.context.action.params(:input)),
      req.input,
      req.context.action.authorize(req, req.user),
      []
    )

    case ret do
      {:error, msg} ->
        {:error, msg, http_status: 403}

      {data, params} ->
        {:ok, %{req | input: data}, params}
    end
  end

  def authorize(%HaveAPI.Response{} = res) do
    ret = do_authorize(
      param_names(res.context.action.params(:output)),
      res.output,
      res.context.action.authorize(res, res.context.user),
      [HaveAPI.Meta.namespace]
    )

    case ret do
      {:error, msg} ->
        {:error, msg, http_status: 403}

      {data, _params} ->
        {:ok, %{res | output: data}}
    end
  end

  defp param_names(nil), do: []
  defp param_names(params), do: Enum.map(params, &(&1.name))

  defp do_authorize(_params, _data, :deny, _ignore), do: {:error, "Access not authorized"}
  defp do_authorize(_params, _data, {:deny, msg}, _ignore), do: {:error, msg}
  defp do_authorize(params, data, :allow, _ignore), do: {data, params}
  defp do_authorize(params, nil, {:allow, _opts}, _ignore), do: {nil, params}

  defp do_authorize(params, data, {:allow, opts}, ignore) when is_list(opts) do
    authorized = authorized_params(params, opts, ignore)
    {filter_data(authorized, data), authorized}
  end

  defp authorized_params(params, opts, ignore) do
    Enum.reduce(
      opts,
      params,
      fn {k,v}, acc ->
        do_authorize_opts(acc, k, v, ignore)
      end
    )
  end

  defp do_authorize_opts(params, :blacklist, list, _ignore) when is_list(list) do
    Enum.filter(params, fn p -> not (p in list) end)
  end

  defp do_authorize_opts(params, :whitelist, list, ignore) when is_list(list) do
    Enum.filter(params, fn p -> not (p in ignore) && p in list end)
  end

  defp filter_data(params, data) do
    data
    |> Enum.filter(fn {k, _v} -> k in params end)
    |> Enum.into(%{})
  end
end
